larpandoracontent/LArThreeDReco/LArCosmicRay/ThreeViewDeltaRayMatchingAlgorithm.cc

```diff
@@ -0,0 +1,164 @@
+{
+    DeltaRayOverlapResult overlapResult;
+    PANDORA_THROW_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, this->CalculateOverlapResult(pClusterU, pClusterV, pClusterW, overlapResult));
+
+    if (overlapResult.IsInitialized())
+        this->GetMatchingControl().GetOverlapTensor().SetOverlapResult(pClusterU, pClusterV, pClusterW, overlapResult);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeViewDeltaRayMatchingAlgorithm::CalculateOverlapResult(const Cluster *const pClusterU, const Cluster *const pClusterV, const Cluster *const pClusterW,
+    DeltaRayOverlapResult &overlapResult) const
+{
+    float chiSquaredSum(0.f);
+    unsigned int nSamplingPoints(0), nMatchedSamplingPoints(0);
+    XOverlap xOverlapObject(0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f);
+    
+    StatusCode statusCode(this->PerformThreeViewMatching(pClusterU, pClusterV, pClusterW, chiSquaredSum, nSamplingPoints, nMatchedSamplingPoints, xOverlapObject));
+
+    if (statusCode != STATUS_CODE_SUCCESS)
+        return statusCode;
+
+    PfoList commonMuonPfoList;
+    this->FindCommonMuonParents(pClusterU, pClusterV, pClusterW, commonMuonPfoList);
+
+    if (commonMuonPfoList.empty())
+        return STATUS_CODE_NOT_FOUND;
+
+    overlapResult = DeltaRayOverlapResult(nMatchedSamplingPoints, nSamplingPoints, chiSquaredSum, xOverlapObject, commonMuonPfoList);
+    
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeViewDeltaRayMatchingAlgorithm::FindCommonMuonParents(const Cluster *const pClusterU, const Cluster *const pClusterV, const Cluster *const pClusterW,
+    PfoList &commonMuonPfoList) const
+{
+    ClusterList consideredClustersU, consideredClustersV, consideredClustersW;
+    PfoList nearbyMuonPfosU, nearbyMuonPfosV, nearbyMuonPfosW;
+    
+    this->GetNearbyMuonPfos(pClusterU, consideredClustersU, nearbyMuonPfosU);
+
+    if (nearbyMuonPfosU.empty())
+        return;
+    
+    this->GetNearbyMuonPfos(pClusterV, consideredClustersV, nearbyMuonPfosV);
+
+    if (nearbyMuonPfosV.empty())
+        return;
+    
+    this->GetNearbyMuonPfos(pClusterW, consideredClustersW, nearbyMuonPfosW);
+
+    if (nearbyMuonPfosW.empty())
+        return;    
+
+    for (const ParticleFlowObject *const pNearbyMuonU : nearbyMuonPfosU)
+    {
+        for (const ParticleFlowObject *const pNearbyMuonV : nearbyMuonPfosV)
+        {
+            for (const ParticleFlowObject *const pNearbyMuonW : nearbyMuonPfosW)
+            {
+                if ((pNearbyMuonU == pNearbyMuonV) && (pNearbyMuonV == pNearbyMuonW))
+                    commonMuonPfoList.push_back(pNearbyMuonU);
+            }
+        }
```

Comment:
There are a couple of options here for improving the efficiency. Under the assumption that a PFO can only appear in each list once you can split the conditional expressions to check U == V inside the V loop and simply continue if there is no match, and then perform a similar V == W check for the actual push_back (you could opt for emplace_back here, which is more efficient given we don't usually need a copy for these operations).

The other option would be to use `std::find(pNearbyMuonV->begin(), pNearbyMuonV->end(), pNearbyMuonU)` to see if it's the same PFO and then only do the second find if it is, avoiding the need for the two inner loops altogether (obviously find is still performing iteration, but I suspect it will be more efficient).